### PR TITLE
[JENKINS-33256] Introducing SCMSource.getTrustedRevision

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-33256](https://issues.jenkins-ci.org/browse/JENKINS-33256)

New API for potential use by branch source plugins.

@reviewbybees esp. @stephenc

### CODE REVIEW

**Comment by stephenc** [comment]
:bee: makes sense... if a little domain specific

File: src/main/java/jenkins/scm/api/SCMSource.java [hunk 1]
```diff
@@ -271,6 +271,34 @@ public final SCM build(@NonNull SCMHead head) {
...
+     * come from principals operating with the same authorization as the principal creating the job,
+     * or at least with authorization to create a similar job.
+     * <p>A source may however collect revisions from untrusted third parties and submit them for builds.
+     * If the project type performing the build loads instructions from the same revision,
+     * this might allow the job to be subverted to perform unauthorized build steps or steal credentials.
+     * <p>By replacing the supplied revision with a trusted variant, a source can defend against such attacks.
+     * It is up to the project type to determine which files should come from a trusted replacement.
+     * Regular project sources should come from the original;
+     * Jenkins-specific scripting commands or configuration should come from the replacement, unless easily sandboxed;
+     * scripts for external build tools should come from the original if possible.
```
Comment (recena) [inline src/main/java/jenkins/scm/api/SCMSource.java]: `newline` would be great to legibility

**Comment by recena** [comment]
:bee:

**Comment by kzantow** [comment]
:bee: :+1: for the docs :)

File: src/main/java/jenkins/scm/api/SCMSource.java (hunk 1 above)
Comment (recena) [inline src/main/java/jenkins/scm/api/SCMSource.java]: Please... ;)

Comment (jglick) [inline src/main/java/jenkins/scm/api/SCMSource.java]: There are plenty of newlines here, I am not sure what is being requested.

Comment (kzantow) [inline src/main/java/jenkins/scm/api/SCMSource.java]: Probably blank lines to separate paragraphs.

**Comment by stephenc** [comment]
:bee:

**Comment by stephenc** [comment]
:+1: as maintainer

**Pull request merged**